src/devices/GrovePi/GrovePi.cs

```diff
@@ -0,0 +1,226 @@
+            byte tries = 0;
+            IOException innerEx = new IOException();
+            while (tries < MaxRetries)
+            {
+                try
+                {
+                    _i2cDevice.Write(outArray);
+                    return;
+                }
+                catch (IOException ex)
+                {
+                    // Give it another try
+                    innerEx = ex;
+                    tries++;
+                    Thread.Sleep(10);                    
+                }
+            }
+
+            throw new IOException($"{nameof(WriteCommand)} error writting command", innerEx);
+        }
+
+        /// <summary>
+        /// Read data from GrovePi
+        /// </summary>
+        /// <param name="commands">The GrovePi command</param>
+        /// <param name="pin">The pin to read</param>
+        /// <returns></returns>
+        public byte[] ReadCommand(GrovePiCommands commands, GrovePort pin)
+        {
+            int numberBytesToRead = 0;
+            switch (commands)
+            {
+                case GrovePiCommands.DigitalRead:
+                    numberBytesToRead = 1;
+                    break;
+                case GrovePiCommands.AnalogRead:
+                case GrovePiCommands.UltrasonicRead:
+                case GrovePiCommands.LetBarGet:
+                    numberBytesToRead = 3;
+                    break;
+                case GrovePiCommands.Version:
+                    numberBytesToRead = 4;
+                    break;
+                case GrovePiCommands.DhtTemp:
+                    numberBytesToRead = 9;
+                    break;
+                // No other commands are for read
+                default:
+                    return null;
+            }
+            byte[] outArray = new byte[numberBytesToRead];
+            byte tries = 0;
+            IOException innerEx = new IOException();
+            while (tries < MaxRetries)
+            {
+                try
+                {
+                    _i2cDevice.Read(outArray);
+                    return outArray;
+                }
+                catch (IOException ex)
+                {
+                    // Give it another try
+                    innerEx = ex;
+                    tries++;
+                    Thread.Sleep(10);
+                }
+            }
+
+            throw new IOException($"{nameof(ReadCommand)} error read command", innerEx);
+        }
+
+        /// <summary>
+        /// Read a digital pin, equivalent of digitalRead on Arduino
+        /// </summary>
+        /// <param name="pin">The GroovePi pin to read</param>
+        /// <returns>Returns the level either High or Low</returns>
+        public PinValue DigitalRead(GrovePort pin)
+        {
+            WriteCommand(GrovePiCommands.DigitalRead, pin, 0, 0);
+            byte tries = 0;
+            IOException innerEx = new IOException();
```

Comment:
Why not IOException `innerEx = null` or if compiler allows just declaration? I don't think this value is ever used

Reply:
yes, it does accept null. The simple declaration was not enough. So I jsut created a new exception which is not needed as null is working.